lib/twom.c

```diff
@@ -0,0 +1,3312 @@
+    // allows us to still re-seek all the way back to this record
+    // and then advance from there if the file gets changed between
+    // starting the cursor and our first read.
+    if (!cur->loc.is_exactmatch || (flags & TWOM_SKIPROOT))
+        cur->started = 1;
+
+done:
+    if (r) twom_cursor_fini(&cur);
+    else *curp = cur;
+
+    return r;
+}
+
+// useful for when you started with an external transaction that you're
+// not finished with, just clean up the cursor
+void twom_cursor_fini(struct twom_cursor **curp)
+{
+    struct twom_cursor *cur = *curp;
+    if (!cur) return;
+    if (cur->loc.file) {
+        cur->loc.file->refcount--;
+        cur->loc.file = NULL;
+    }
+    free(cur);
+    *curp = NULL;
+    return;
+}
+
+/* foreach allows for subsidiary operations in 'cb',
+ * if there's a transaction then cb must use it.
+ * If cb returns a non-zero result, then it gets
+ * returned to the caller.  NOTE the values -128 to +1
+ * are reserved.  This is a sucky interface really,
+ * there should be another way to return errors.
+ *
+ * If you really need to return anything more complex, you
+ * can store the return value inside 'rock' instead.
+*/
+int twom_txn_foreach(struct twom_txn *txn,
+                     const char *prefix, size_t prefixlen,
+                     twom_cb *goodp, twom_cb *cb, void *rock,
+                     int flags)
+{
+    int r = 0, cb_r = 0;
+    const char *key = NULL;
+    size_t keylen = 0;
+    const char *data = NULL;
+    size_t datalen = 0;
+    struct twom_cursor *cur = NULL;
+
+    assert(cb);
+    if (prefixlen) assert(prefix);
+
+    r = twom_txn_begin_cursor(txn, prefix, prefixlen, &cur, flags);
+    if (r) goto done;
+
+    while ((r = twom_cursor_next(cur, &key, &keylen, &data, &datalen)) == 0) {
+        /* does it match prefix? */
+        if (prefixlen) {
+            if (keylen < prefixlen) break;
+            if (txn->file->compar(key, prefixlen, prefix, prefixlen)) break;
+        }
+
+        if ((!goodp || goodp(rock, key, keylen, data, datalen))) {
+            /* make callback */
+            cb_r = cb(rock, key, keylen, data, datalen);
+            if (cb_r) break;
+        }
+    }
+
+    // safely finished
+    if (r == TWOM_DONE) r = 0;
+
+ done:
+    twom_cursor_fini(&cur);
+
+    return r ? r : cb_r;
+}
+
+int twom_txn_store(struct twom_txn *txn,
+                   const char *key, size_t keylen,
+                   const char *data, size_t datalen,
+                   int flags)
+{
+    /* no writing a readonly database */
+    if (txn->db->readonly)
+        return TWOM_READONLY;
+
+    assert (txn == txn->db->write_txn);
+    assert(key && keylen);
+
+    return skipwrite(txn, key, keylen, data, datalen, flags);
+}
+
+const char *twom_db_fname(struct twom_db *db)
+{
+    return db->fname;
+}
+
+int twom_db_sync(struct twom_db *db)
+{
+    if (!db->openfile) return 0;
+    return tm_commit(db, db->openfile->size);
+}
+
+size_t twom_db_generation(struct twom_db *db)
```

Comment:
This function is not called from anywhere.

Reply:
Likewise